shenyu-admin/src/main/java/org/apache/shenyu/admin/service/impl/ApiServiceImpl.java

```diff
@@ -0,0 +1,142 @@
+        this.apiMapper = apiMapper;
+        this.tagRelationMapper = tagRelationMapper;
+
+    }
+
+    @Override
+    @Transactional(rollbackFor = Exception.class)
+    public String createOrUpdate(final ApiDTO apiDTO) {
+        return StringUtils.isBlank(apiDTO.getId()) ? this.create(apiDTO) : this.update(apiDTO);
+    }
+
+    /**
+     * update.
+     *
+     * @param apiDTO apiDTO
+     * @return update message
+     */
+    private String update(final ApiDTO apiDTO) {
+        ApiDO apiDO = ApiDO.buildApiDO(apiDTO);
+        final int updateRows = apiMapper.updateByPrimaryKeySelective(apiDO);
+        if (CollectionUtils.isNotEmpty(apiDTO.getTagIds()) && updateRows > 0) {
+            List<String> tagIds = apiDTO.getTagIds();
+            List<TagRelationDO> tags = Lists.newArrayList();
+            Timestamp currentTime = new Timestamp(System.currentTimeMillis());
+            for (String tagId : tagIds) {
+                tags.add(TagRelationDO.builder()
+                        .id(UUIDUtils.getInstance().generateShortUuid())
+                        .apiId(apiDO.getId())
+                        .tagId(tagId)
+                        .dateCreated(currentTime)
+                        .dateUpdated(currentTime)
+                        .build());
+            }
+            tagRelationMapper.deleteByApiId(apiDO.getId());
+            tagRelationMapper.batchInsert(tags);
+        }
+        return ShenyuResultMessage.UPDATE_SUCCESS;
+    }
+
+    /**
+     * create.
+     *
+     * @param apiDTO apiDTO
+     * @return create message
+     */
+    private String create(final ApiDTO apiDTO) {
+        ApiDO apiDO = ApiDO.buildApiDO(apiDTO);
+        final int insertRows = apiMapper.insertSelective(apiDO);
+        if (CollectionUtils.isNotEmpty(apiDTO.getTagIds()) && insertRows > 0) {
+            List<String> tagIds = apiDTO.getTagIds();
+            List<TagRelationDO> tags = Lists.newArrayList();
+            Timestamp currentTime = new Timestamp(System.currentTimeMillis());
+            for (String tagId : tagIds) {
+                tags.add(TagRelationDO.builder()
+                        .id(UUIDUtils.getInstance().generateShortUuid())
+                        .apiId(apiDO.getId())
+                        .tagId(tagId)
+                        .dateCreated(currentTime)
+                        .dateUpdated(currentTime)
+                        .build());
+            }
+            tagRelationMapper.batchInsert(tags);
+        }
+        return ShenyuResultMessage.CREATE_SUCCESS;
+    }
+
+    @Override
+    @Transactional(rollbackFor = Exception.class)
+    public String delete(final List<String> ids) {
+        // select api id.
+        List<ApiDO> apis = this.apiMapper.selectByIds(ids);
+        if (CollectionUtils.isEmpty(apis)) {
+            return AdminConstants.SYS_API_ID_NOT_EXIST;
+        }
+        // delete apis.
+        this.apiMapper.deleteByIds(ListUtil.map(apis, ApiDO::getId));
+        return StringUtils.EMPTY;
```

Comment:
maybe lost `tagRelationMapper.deleteByApiId` ?

Reply:
Not required for batch deletion?